src/main/java/ma/hibernate/dao/PhoneDaoImpl.java

```diff
@@ -12,11 +20,49 @@ public PhoneDaoImpl(SessionFactory sessionFactory) {
 
     @Override
     public Phone create(Phone phone) {
-        return null;
+        Session session = null;
+        Transaction transaction = null;
+        try {
+            session = factory.openSession();
+            transaction = session.beginTransaction();
+            session.persist(phone);
+            transaction.commit();
+            return phone;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new RuntimeException("Can't create phone " + phone, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
     }
 
     @Override
-    public List<Phone> findAll(Map<String, String[]> params) {
-        return null;
+    public List<Phone> findAll(Map<String, String[]> parameters) {
+        try (Session session = factory.openSession()) {
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<Phone> query = criteriaBuilder.createQuery(Phone.class);
+            Root<Phone> phoneRoot = query.from(Phone.class);
+            List<Predicate> predicate = new ArrayList<>();
+            for (Map.Entry<String, String[]> entry : parameters.entrySet()) {
+                CriteriaBuilder.In<String> inPredicate =
+                        criteriaBuilder.in(phoneRoot.get(entry.getKey()));
+                List<String> stringValues = Arrays.asList(entry.getValue());
+                for (String parameter : stringValues) {
+                    inPredicate.value(parameter);
+                }
+                Predicate tempPredicate = criteriaBuilder.and(inPredicate);
+                predicate.add(tempPredicate);
```

Comment:
```suggestion
                predicates.add(phoneRoot.get(entry.getKey())
                        .in(entry.getValue()));
```
what do you think? this will make the code much shorter

Reply:
Wow. O_o Thank you. It's much shorter. Thank you. =)